SCClassLibrary/Common/Core/CondVar.sc

```diff
@@ -0,0 +1,116 @@
+CondVar {
+	var waitingThreads;
+
+	*new { ^super.newCopyArgs(Array.new, Array.new) }
+
+	wait { |predicate|
+		if(predicate.isNil) {
+			this.prWait
+		} {
+			// If the predicate is already true, we return immediately. If not,
+			// we block until the predicate is true after being signalled.
+			while { predicate.value.not } { this.prWait }
+		}
+	}
+
+	waitFor { |timeoutBeats, predicate|
+		var endingBeats;
+
+		// Convert timeoutBeats to a value we know is safe to call `wait` on,
+		// since an exception thrown in the timeout thread is a very bad thing.
+		// May throw if preconditions are not met.
+		timeoutBeats = this.prHandleTimeoutPreconditions(timeoutBeats);
+
+		// This is a slightly more complex version of the body of `wait`. The
+		// extra complexity comes from maintaining the timeout requirement.
+		if(predicate.isNil) {
+			// Interface requirement: immediately return if timeout is non-positive
+			if(timeoutBeats <= 0) { ^false };
+			^this.prWaitFor(timeoutBeats)
+		} {
+			endingBeats = thisThread.beats + timeoutBeats;
+
+			// If the predicate is already true, we return true immediately.
+			while {
+				predicate.value.not
+			} {
+				// We only get here when the predicate is not true, so we can return
+				// false immediately if we are timed out.
+				if(thisThread.beats >= endingBeats) { ^false };
+				// If the timeout expires, return the value of the predicate; otherwise,
+				// we were signalled, so go back through the loop.
+				if(this.prWaitFor(endingBeats - thisThread.beats).not) { ^predicate.value };
+			};
+
+			^true
+		}
+	}
+
+	signalOne {
+		if (waitingThreads.isEmpty.not) {
+			this.prWakeThread(waitingThreads.removeAt(0));
+		}
+	}
+
+	signalAll {
+		waitingThreads.do(this.prWakeThread(_));
+		waitingThreads = Array.new;
+	}
+
+	shallowCopy { this.shouldNotImplement(thisMethod) }
+	deepCopy { this.shouldNotImplement(thisMethod) }
+
+	prWait {
+		this.prSleepThread(thisThread);
+		\wait.yield
+	}
+
+	// Returns true iff we were woken via signal (and not timeout)
+	prWaitFor { |timeoutBeats|
+		// precondition: timeoutBeats is a Float or Integer, positive, and not inf
+		var waitingThread = thisThread;
+		var didNotTimeout = true;
+		var timeoutThread = Routine {
+			var wokenThread;
+
+			timeoutBeats.wait;
+
+			didNotTimeout = false;
+
+			// wokenThread must be non-nil. If the thread woke before this timeout fired,
```

Comment:
"If the condition was signalled before this timeout fired, this routine will be stopped (below, after prWait) and we won't get here" -- I didn't get *how* the thread would awake (it should never wake without a signal, right?).

Reply:
>  (it should never wake without a signal, right?).

correct, you could still reschedule it yourself I guess, but I don't think that would have any effect beyond what signal does. I haven't thought through that very carefully, so I could be wrong. in my view, not something to protect against, but I can think it through more if you feel differently.